llvm/lib/SYCLLowerIR/MutatePrintfAddrspace.cpp

```diff
@@ -0,0 +1,230 @@
+//
+// A transformation pass which detects non-constant address space
+// literals usage for the first argument of SYCL experimental printf
+// function, and moves the string literal to constant address
+// space. This a temporary solution for printf's support of generic
+// address space literals; the pass should be dropped once SYCL device
+// backends learn to handle the generic address-spaced argument properly.
+//===----------------------------------------------------------------------===//
+
+#include "llvm/SYCLLowerIR/MutatePrintfAddrspace.h"
+
+#include "llvm/Analysis/ValueTracking.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/InitializePasses.h"
+
+using namespace llvm;
+
+namespace {
+// Wrapper for the pass to make it working with the old pass manager
+class SYCLMutatePrintfAddrspaceLegacyPass : public ModulePass {
+public:
+  static char ID;
+  SYCLMutatePrintfAddrspaceLegacyPass() : ModulePass(ID) {
+    initializeSYCLMutatePrintfAddrspaceLegacyPassPass(
+        *PassRegistry::getPassRegistry());
+  }
+
+  // run the SYCLMutatePrintfAddrspace pass on the specified module
+  bool runOnModule(Module &M) override {
+    ModuleAnalysisManager MAM;
+    auto PA = Impl.run(M, MAM);
+    return !PA.areAllPreserved();
+  }
+
+private:
+  SYCLMutatePrintfAddrspacePass Impl;
+};
+
+struct CallReplacer;
+class AddrspaceReplacer {
+public:
+  static constexpr unsigned ConstantAddrspaceID = 2;
+
+  AddrspaceReplacer() = delete;
+  AddrspaceReplacer(Module *M) : M(M) {
+    Type *Int8Ty = Type::getInt8Ty(M->getContext());
+    CASLiteralType = PointerType::get(Int8Ty, ConstantAddrspaceID);
+  }
+  ~AddrspaceReplacer() {
+    for (Function *F : FunctionsToDrop)
+      F->eraseFromParent();
+  }
+  size_t getReplacedCallsCount() { return ReplacedCallsCount; }
+
+  void runOnFunctionDeclaration(Function *F);
+
+private:
+  Module *M;
+  PointerType *CASLiteralType;
+  size_t ReplacedCallsCount = 0;
+  /// If the variadic version gets picked during FE compilation, we'll only have
+  /// 1 function to replace. However, unique declarations are emitted for each
+  /// of the non-variadic (variadic template) calls.
+  SmallVector<Function *, 8> FunctionsToDrop;
+
+  Function *getCASPrintfFunction(Function *GenericASPrintfFunc);
+  Constant *getCASLiteral(GlobalVariable *GenericASLiteral);
+};
+} // namespace
+
+char SYCLMutatePrintfAddrspaceLegacyPass::ID = 0;
+INITIALIZE_PASS(SYCLMutatePrintfAddrspaceLegacyPass,
+                "SYCLMutatePrintfAddrspace",
+                "Move SYCL printf literal arguments to constant address space",
+                false, false)
+
+// Public interface to the SYCLMutatePrintfAddrspacePass.
+ModulePass *llvm::createSYCLMutatePrintfAddrspaceLegacyPass() {
+  return new SYCLMutatePrintfAddrspaceLegacyPass();
+}
+
+PreservedAnalyses
+SYCLMutatePrintfAddrspacePass::run(Module &M, ModuleAnalysisManager &MAM) {
+  AddrspaceReplacer ASReplacer(&M);
```

Comment:
What `AddrspaceReplacer` allows for is to store module context-specific information (`Module *`, `CASLiteralType`, `FunctionsToDrop`, etc.) as class members instead of `SYCLMutatePrintfAddrspacePass::runOnModule()`-scope variables. Personally, I find this important for readability purposes - when skimming through the pass body, I'm driven to the high-level algorithmic nature of the pass structure and prefer not to be distracted by the helper variables & structures.
The alternative could be to store these Module-specific variable values in the pass class itself, but then we'd have to reassign values to them within each `runOnModule()` call. Perhaps a helper function could do the assignment and retain the cleanliness. Is the latter a more comprehensive approach compared to the separate `AddrspaceReplacer` entity?

Reply:
> The alternative could be to store these Module-specific variable values in the pass class itself

That is how it is usually done, i.e. it is not necessary to more al the logic and data into a separate class to make `runOnModule` function clean - you can organize all the logic and data within `SYCLMutatePrintfAddrspacePass` and achieve the same level of cleanliness.

> but then we'd have to reassign values to them within each runOnModule() call. Perhaps a helper function could do the assignment and retain the cleanliness.

We have only a couple of assignments, so I don't think that it would affect cleanliness much. `Module` shouldn't be even stored anywhere, because it is easily accessible though a `Function`, for example. `getCASPrintfFunction` which uses the module can only be called once, because we expect the replacement to be the same for all `printf` calls. `FunctionsToDrop` do not require any assignment/special handling, because we expect the vector to be empty at the end of `runOnModule` function so it is ready to be used again if it is called for the second time. `CASLiteralType` - that is just a one-liner (almost).

Reply:
Done along the lines of our discussion here & offline - d2fb5f9baba95396d6af6ac419d296f1dc0161a7 focuses on code re-design specifically